### PR TITLE
Declare non-const enums if preserveConstEnums is true

### LINKED ISSUE #37774: isolatedModules doesn't respect enabled preserveConstEnum option what the project might be build with
## Search Terms

isolatedModules, preserveConstEnum, const enum

## Suggestion

The compiler should respect/understand that the package/module was compiled with enabled `preserveConstEnums` compiler option.

## Use Cases

Let's say we have a project written in TypeScript, which uses const enums.

The project uses them because the main `const enum`'s advantage is inlining values instead of runtime accessing to the value.

In other hand the maintainers understand that const enums compiles in nothing in JS code, and that means that consumers can't use a const enums in their JS code (in some cases in TS code either, see above). Thus, they decided to enable `preserveConstEnums` compiler option to leave all const enums in JS code and allow consumers use them as well.

Up to this moment everything is good. But the next who appears on the stage is `isolatedModules` (say hello to `create-react-app`, webpack's transpile-only mode, `transpileModule` compiler's API and so on). This compiler option is used to detect that imported module could be transpiler without compilation into JS code and everything will work well. One of checks for this option is that you don't import const enums (error `TS2748: Cannot access ambient const enums when the '--isolatedModules' flag is provided`).

To summarize, if the project/package is compiled with enabled `preserveConstEnums`, in `isolatedModules` mode the compiler shouldn't fail with error about unavailability of using const enums, because that's not actually true.

## Examples

Quite possible we need to provide that information to the compiler somehow (by providing tsconfig.export.json or in package.json for instance), but I don't even know how it would be done.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by jablko**
Currently if you `tsc --declaration --isolatedModules`, the compiler turns non-ambient const enums (`.ts` files) into ambient const enums (`.d.ts` files) that can't themselves be consumed with `--isolatedModules`: [Ambient const enums (values -- types are fine) are incompatible with `isolatedModules`](https://www.typescriptlang.org/tsconfig#references-to-const-enum-members).

In every other way `tsc --isolatedModules` ignores the const modifier -- treats const enums as non-const:
- Emits `.js` values (like `preserveConstEnums`)
- Doesn't inline values and doesn't elide imports (*unlike* `preserveConstEnums`)

This PR doesn't emit the const modifier if `isolatedModules` is true, effectively suppressing the const modifier entirely in that case. `tsc --declaration --isolatedModules` then turns *non-ambient* const enums into ambient *non-const* enums which can in turn be consumed with `--isolatedModules`: https://github.com/microsoft/TypeScript/issues/37774#issuecomment-608570356

The affected scenario is when you:
1. Produce `.d.ts` with `isolatedModules: true` (currently: const modifiers preserved, this PR: const modifiers dropped)
2. Consume the produced `.d.ts`:
   - With `isolatedModules: true`: Currently: error. This PR: no error, values and imports preserved, same as all enums, (non-ambient) const enums included.
   - With `isolatedModules: false`: Currently: values inlined, imports elided. This PR: same as with `isolatedModules: true` (values and imports preserved).

~When a `.d.ts` is produced with `isolatedModules: true` and consumed with `isolatedModules: false`, what's the point of inlining values and eliding imports? If you consume any part of the produced library that itself uses the const enum values, you'll end up importing the enum's `.js` values.~

~So this PR will only introduce imports that don't currently exist if you publish a library that declares const enum values it doesn't itself use, and in that case if you *really* want to publish isolatedModules-incompatible ambient const enums, produce its `.d.ts` with `isolatedModules: false`.~

Edit: Inlining enums from other projects [comes with pitfalls](https://github.com/microsoft/TypeScript/issues/37774#issuecomment-961392924):
> 2. Consumers can inline enum values from upstream version A at compile time and import version B at runtime, with different values, if you're not very careful:
>    - https://github.com/microsoft/TypeScript/issues/37179#issuecomment-594070333
>    - https://github.com/microsoft/TypeScript/pull/5297#issuecomment-151634916
>    - https://github.com/microsoft/TypeScript/issues/5219#issue-110947903

Roughly this PR propagates today's isolatedModules const enum behavior transitively to `.d.ts` consumers.

Fixes #37774

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -3261,7 +3262,17 @@ namespace ts {
...
 
         function emitEnumDeclaration(node: EnumDeclaration) {
-            emitModifiers(node, node.modifiers);
+            // Elide const modifiers if isolatedModules is true
+            emitModifiers(
+                node,
+                factory.createNodeArray(
+                    filter(
+                        node.modifiers,
+                        modifier => !printerOptions.isolatedModules || modifier.kind !== SyntaxKind.ConstKeyword
```
Comment (jablko) [inline src/compiler/emitter.ts:3271]: Alternatively this could be done when parsing vs. emitting const enums?

**Comment by jablko** [comment]
This doesn't fix #37774 if `isolatedModules: false, preserveConstEnums: true`. The advantage of publishing with `isolatedModules: false, preserveConstEnums: true` is that you can consume parts of the produced library that themselves use const enum values, without importing the enum's `.js` values ...

**Comment by sandersn** [comment]
@jablko is this a proof of concept? When I skimmed the issue I didn't see that it had arrived at a single solution yet.

**Comment by jablko** [comment]
It's a proposed solution: https://github.com/microsoft/TypeScript/issues/37774#issuecomment-961392924, but you're right: It's awaiting feedback (issue and proposal).

I think it's right because:
1. Publishing ambient const enums is a current problem:
   - Create React App enforces `isolatedModules: true` ([reportedly](https://github.com/microsoft/TypeScript/issues/37774#issuecomment-720624134)?)
   - https://github.com/microsoft/TypeScript/issues/37179#issue-574347057
2. Deconstifying `.d.ts` files [in a build step](https://github.com/microsoft/TypeScript/blob/1a981d1df1810c868a66b3828497f049a944951c/Gulpfile.js#L144) requires author buyin. The only other mitigation today is to avoid const enums entirely, which has performance implications (though most often small?).
3. While two out of three ambient const enum problems have proposed remedies (a [`preserved` modifier](https://github.com/microsoft/TypeScript/issues/37774#issuecomment-902988750) would address `isolatedModules` incompatibility, https://github.com/microsoft/TypeScript/pull/46556#issue-1037783024 addresses import elision), there remains a [compile time vs. runtime pitfall](https://github.com/microsoft/TypeScript/issues/37774#issuecomment-961392924) to inlining inter-project enums (ambient const enums, roughly).

&therefore; `tsc --declaration --preserveConstEnums` should automatically not transform const enums -> ambient const enums (deconstify ambient enums), I think.
- The cases where you *want* to inline inter-project enums (with consequent pitfall) are already addressed by `preserveConstEnums: false`.
- The cases where you want some consumers to inline enums and others not to are:
  1. [More confusing than useful](https://github.com/microsoft/TypeScript/issues/37774#issuecomment-902988750)?
  2. Doable anyway in a build step, if you know what you're doing (like deconstification is today)?

The difference between this PR and a `preserved` modifier is that the modifier would presumably inline preserved inter-project const enums if `isolatedModules` is false. This PR is simpler (no new kind of enum) and desirable in the majority of and the entry-level cases (avoids pitfall, little or no downside?).

**Comment by sandersn** [comment]
Based on the discussion at the December 1st design meeting, we aren't convinced that improving const enums is worthwhile, since people should be using the feature *less* over time rather than more.

**Comment by jablko** [comment]
All of the const enum pitfalls remain if you *don't* use `preserveConstEnums: true`, which should discourage people (as much as today)? :smirk:

- Avoiding const enums is preferable, but the number of people who should truly use them isn't zero.
- In those few cases, the current workaround (a [build step](https://github.com/microsoft/TypeScript/blob/1a981d1df1810c868a66b3828497f049a944951c/Gulpfile.js#L144)) isn't onerous, but this change is commensurately small, and strictly positive, I think?

**Comment by jablko** [comment]
I've added a commit to inline plain enums from [referenced projects](https://www.typescriptlang.org/docs/handbook/project-references.html).

Currently you get different behavior consuming .ts files and consuming the .d.ts outputs of those same files: const enums in .ts files are (ordinarily) compatible with `--isolatedModules` (because it flips on `--preserveConstEnums`), whereas the .d.ts outputs aren't (ambient const enums are incompatible with single-file transpilation).

This came up again [here](https://github.com/microsoft/TypeScript/issues/48040#issuecomment-1055685379). An example is you can't use const enums from referenced projects with `--isolatedModules`. You can use const enums from .ts files, but the CLI uses the .d.ts outputs for referenced projects. That error shows up in the CLI but not in the editor, which defaults to using .ts files.

This PR fixes that, but now you get different behavior between .ts files and their .d.ts outputs when built with `--preserveConstEnums` and consumed *without* `--isolatedModules`: const enums in .ts files are inlined whereas their plain enum .d.ts outputs aren't. The latest commit addresses that.

- *With* `--isolatedModules`, `--preserveConstEnums` and deconstification solve single-file transpilation compatibility by emitting plain enums. `--isolatedModules` already treats const enums as plain enums anyway (doesn't inline).
- *Without* `--isolatedModules` it solves other problems with publishing ambient const enums: https://github.com/microsoft/TypeScript/issues/37179#issuecomment-594070333 (JS consumers, compile-time vs. runtime version compatibility).

Not inlining enums from referenced projects could be a problem, however. The latest commit solves that by inlining plain enums. Inlining *all* enums would open up compile-time vs. runtime version compatibility with published .d.ts outputs, but inlining plain enums from referenced projects is safe because they're part of the same build.

The logic is that we inline const enums from .ts files, we use deconstified .d.ts outputs for referenced projects, so inline plain enums from referenced projects.

**Comment by sandersn** [comment]
To help with PR housekeeping, I'm going to close this PR while it's still waiting on its bug to be accepted.

**Pull request closed without merging**